File: blf-ts/src/chunks/haloreach/v11860_10_07_24_0147_omaha_release/s_blf_chunk_network_lsp_heartbeat_data.ts
import { c } from "@craftycodie/cstruct";
import {
  s_network_lsp_heartbeat_player_data,
  s_network_lsp_heartbeat_session_data,
} from "../../../blam/haloreach/v12065_11_08_24_1738_tu1actual/networking/logic/network_lsp_presence";
import { blf, CStructBLFChunk } from "../../../blf_chunk";

/** Reach release LSP presence heartbeat upload (`phbt` 5.1, 0x1BB-byte body). */
@blf.chunk("phbt", 5.1)
@c.struct()
export class s_blf_chunk_network_lsp_heartbeat_data extends CStructBLFChunk {
  @c.field("u8")
  has_players = 0;

  @c.field("u8")
  local_player_count = 0;

  @c.field(s_network_lsp_heartbeat_player_data, { count: 4 })
  players = Array.from(
    { length: 4 },
    () => new s_network_lsp_heartbeat_player_data()
  );

  @c.field("u64")
  machine_id = 0n;

  @c.field("u8", { count: 8 })
  unknowneeA = Array.from({ length: 8 }, () => 0);

  @c.field(s_network_lsp_heartbeat_session_data)
  session_data = new s_network_lsp_heartbeat_session_data();
}

File: blf_lib/src/blf/chunks/haloreach/v12065_11_08_24_1738_tu1actual/s_blf_chunk_network_lsp_heartbeat_data.rs
use binrw::binrw;
use blf_lib::types::array::StaticArray;
use blf_lib::types::u64::Unsigned64;
use serde::{Deserialize, Serialize};
use blf_lib_derivable::blf::chunks::BlfChunkHooks;
use blf_lib_derive::BlfChunk;

pub use crate::blam::haloreach::v12065_11_08_24_1738_tu1actual::networking::logic::network_lsp_presence::{
    s_network_lsp_heartbeat_player_data, s_network_lsp_heartbeat_session_data,
};

/// Reach TU1 LSP presence heartbeat upload (`phbt` 5.1, 0x1BB-byte body).
#[binrw]
#[derive(BlfChunk, Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
#[Header("phbt", 5.1)]
#[Size(0x1BB)]
#[brw(big)]
pub struct s_blf_chunk_network_lsp_heartbeat_data {
    pub has_players: u8,
    pub local_player_count: u8,
    pub players: StaticArray<s_network_lsp_heartbeat_player_data, 4>,
    pub machine_id: Unsigned64,
    pub unknowneeA: StaticArray<u8, 8>,
    pub session_data: s_network_lsp_heartbeat_session_data,
}

impl BlfChunkHooks for s_blf_chunk_network_lsp_heartbeat_data {}
